primo2/inference/dynamic.py

```diff
@@ -0,0 +1,305 @@
+            be simple strings of the given values, or a np.array specifying
+            the strength of the evidence for each outcome, e.g.: For the binary
+            evidence node E `{"E": "True"}` is equivalent to
+            `{"E": np.array([1.0, 0.0])}`.
+        """
+        r = self._set_evidence(evidence)
+        self._evidence = True
+        return r
+
+    def unroll(self, evidence=None, **kwargs):
+        """Unroll the network, and proceed in time.
+
+        If no evidence has been set (via `set_evidence`) or is specified in the
+        arguments it proceeds without evidence. If evidence has been set
+        previously and no evidence is specified in the arguments it uses this
+        pre-specified evidence. If evidence is set in the arguments, this new
+        evidence is used.
+
+        Calls the method `_unroll`, which does the actual work.
+
+        evidence : dict, optional
+            Dictionary containing the given evidence. The keys represent the
+            variable names of the given evidence and the values can either
+            be simple strings of the given values, or a np.array specifying
+            the strength of the evidence for each outcome, e.g.: For the binary
+            evidence node E `{"E": "True"}` is equivalent to
+            `{"E": np.array([1.0, 0.0])}`.
+        **kwargs
+            Takes the same keyword arguments as `set_evidence` and
+            `_set_evidence`.
+        """
+        if not self._evidence or evidence is not None:
+            self._set_evidence(evidence, **kwargs)
+        r = self._unroll()
+        self._evidence = False
+        return r
+
+    @abc.abstractmethod
+    def _set_evidence(self, evidence=None, **kwargs):
+        """Set evidence for the network (doing the actual work). (abstract)
+
+        See `set_evidence`.
+        """
+        pass
+
+    @abc.abstractmethod
+    def _unroll(self):
+        """Unroll the network (doing the actual work). (abstract)
+
+        See `unroll`.
+        """
+        pass
+
+    @property
+    def t(self):
+        """Get number current time-slice.
+
+        Returns
+        -------
+        int
+            The number of the current time-slice.
+        """
+        return self._t
+
+
+class PriorFeedbackExact(DBNInferenceMethod):
+    """Prior feedback-based approach to unrolling a DBN, using exact inference.
+
+    Inference is exact and based on factor trees. The marginal posterior
+    probabilities of the variables of the current time-slice are used as prior
+    probabilties for the next time-slice. They are set as the conditional
+    probability distribution of the nodes/variables.
+    """
+
+    def __init__(self, dbn):
+        super(PriorFeedbackExact, self).__init__(dbn)
+        self._dbn = copy.deepcopy(dbn)
+        self._ft = None
+
+    def marginal_probabilities(self, variable):
+        """Get the marginal probabilities of a variable.
+
+        See `DBNInferenceMethod.marginal_probabilities`.
```

Comment:
I am split with on the "See xy". On the one hand side I am completely with you that duplicating the docstring is a bad idea as it only produces more work as well as provide the potential of inconsistent docstrings. On the other hand side, introspection and IDE-Support would fail with these docstrings, as the IDE would pick up, inspect and display these short versions and most likely will not go up to the abstract class. PEP does not talk about abstract classes and only states that subclasses should explicitly either use "override" or "extend" depending on whether or not the function of the superclass is used or not (https://www.python.org/dev/peps/pep-0257/). I guess we can leave it like this for now, but we should maybe think about how we want to handle this in general.